### PR TITLE
Pull Request for Issue2164: PGM Components - Shutters and Valves

### PR DESCRIPTION
**Pull request by dleacock**
Added in basic shutters and valves. 

#2164

### CODE REVIEW

**Comment by dleacock** [comment]
**Working on shutters first:**

Will attempt to create a BeamStatus button bar, similar to BioXAS.  First iteration will feature 4 buttons, one for each shutter. From my understanding of the BeamStatus it is a read only widget. Which suits my purpose. 

- [ ] PGMBeamStatusBar and appropriate classes
- [ ] Add shutters to PGMBeamStatusBar

**Comment by XuetaoLiu-CLS** [comment]
This will be reviewed after #2352 is merged in

**Comment by XuetaoLiu-CLS** [comment]
@dleacock This issue is basically done. I added the shutters and valves components.

I will create a new issue for Beam On/Off actions, which will operate on the Shutters and Valves

**Comment by XuetaoLiu-CLS** [comment]
@helfrij can you reivew this? Thanks

**Comment by XuetaoLiu-CLS** [comment]
@helfrij this should be reviewed after #2352 is merged in.

**Comment by XuetaoLiu-CLS** [comment]
@helfrij can u look at this? thx

File: source/application/BioXAS/BioXASAppController.cpp [hunk 1]
```diff
@@ -188,17 +188,17 @@ void BioXASAppController::onRegionOfInterestBoundingRangeChanged(AMRegionOfInter
 		xasConfiguration_->setRegionOfInterestBoundingRange(region);
 }
 
-void BioXASAppController::goToBeamStatusView(AMControl *control)
+void BioXASAppController::goToBeamlineStatusView(AMControl *control)
 {
-	if (beamStatusView_) {
+	if (beamlineStatusView_) {
```
Comment (helfrij) [inline source/application/BioXAS/BioXASAppController.cpp]: It's a pedantic detail. But the 'beam status' is different than the 'beamline status', which is something we may implement later. Is it possible to keep the original names here?

File: source/beamline/AMEnumeratedControl.h [hunk 2]
```diff
@@ -15,7 +15,7 @@ class AMEnumeratedControl : public AMPseudoMotorControl
 	enum DefaultOptions { Unknown = -1, Invalid = -2 };
 
 	/// Constructor.
-	explicit AMEnumeratedControl(const QString &name, const QString &units, QObject *parent = 0);
+	explicit AMEnumeratedControl(const QString &name, const QString &units, QObject *parent = 0, const QString &description = "");
```
Comment (helfrij) [inline source/beamline/AMEnumeratedControl.h]: 👍

File: source/beamline/PGM/PGMBeamline.cpp [hunk 3]
```diff
@@ -131,6 +131,34 @@ void PGMBeamline::setupMono()
...
 void PGMBeamline::setupComponents()
 {
+	// ==== set up the beamline status
+	// Photon Shutters
+	AMControl *photonShutter1 = new AMReadOnlyPVControl("PhotonShutter1", "PSH1411-I00-01:state", this, "Photon Shutter 1");
+	AMControl *photonShutter2 = new CLSExclusiveStatesControl("PhotonShutter2", "PSH1411-I00-02:state", "PSH1411-I00-02:opr:open", "PSH1411-I00-02:opr:close", this, "Photon Shutter 2");
+	AMControl *photonShutter3 = new CLSExclusiveStatesControl("PhotonShutter3", "PSH1611-3-I20-01:state", "PSH1611-3-I20-01:opr:open", "PSH1611-3-I20-01:opr:close", this, "Photon Shutter 3");
+	AMControl *safetyShutter1 = new CLSExclusiveStatesControl("SafetyShutter1", "SSH1411-I00-01:state",  "SSH1411-I00-01:opr:open", "SSH1411-I00-01:opr:close", this, "Safety Shutter 1");
+
+	beamlineShutters_ = new CLSShutters(QString("PGM Shutters"), this);
```
Comment (helfrij) [inline source/beamline/PGM/PGMBeamline.cpp]: The shutters may be tricky to have as a general CLS class here, if they implement opening and closing of the shutters. Certain shutters need to be opened/closed in a certain order, and that may be different from beamline to beamline. Just checking, is this ordering okay for PGM?

**Comment by helfrij** [comment]
Couple of small questions.

Time Estimate: 5 minutes.

File: source/beamline/PGM/PGMBeamline.cpp (hunk 3 above)
Comment (XuetaoLiu-CLS) [inline source/beamline/PGM/PGMBeamline.cpp]: @helfrij Yes, the beam on/off order are different from beamline to beamline. This CLSShutters/CLSValves/CLSBeamlineStatus classes might be refactored when working on #2355, which will try to create a generic ways to figure out the order.

**Pull request merged**